### ISSUE TITLE
getOrderShippingCostExternal seems not work

### ISSUE THREAD

**Issue by BenTen** [PrestaShop/PrestaShop#36294]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

For making a carrier module with an external api we have to use `getOrderShippingCostExternal()` as wrote in the doc:
https://devdocs.prestashop-project.org/8/modules/carrier/

> **Computing the shipping price**
> 
> To compute the shipping price, PrestaShop needs to call the module’s `getOrderShippingCost()` or `getOrderShippingCostExternal()`.
> 
> The returned value must be a float or false if the carrier is not known or disabled.

But when we set `need_range = 0` and `is_module=1` and then use `getOrderShippingCostExternal()`, the carrier is not display in the list on the FO.

As soon as we set `need_range = 1` and then use `getOrderShippingCost()`, the carrier appear in the list on the FO.

Some issues have been written in last 4 years, but all have been closed due to no answer from reporters.

- https://github.com/PrestaShop/PrestaShop/issues/24591
- https://github.com/PrestaShop/PrestaShop/issues/17803

### Expected behavior

We have to set a range of 0~8000 kg and set 0 in all zones, just to be able to access external api data.

By using an external api to get carrier fee, we don't want to use ranges or zones on PS, we just want communicate with the external API.

### Steps to reproduce

The step to reproduce has been explained by @mehdi-ghezal in an other closed issue.
https://github.com/PrestaShop/PrestaShop/issues/17803#issuecomment-1467372217

### PrestaShop version(s) where the bug happened

1.7.5

**Comment by paulnoelcholot** [comment]
Hello @BenTen,

Thanks for reporting this issue!
Ping @PrestaShop/prestashop-core-developers : Could someone please try to reproduce the issue, it's too technical!

Thanks in advance!

**Comment by Hlavtox** [comment]
@kpodemski I am developing my first ever carrier module and immediately stumbled upon this issue. How come more people are not complaining when the method is just never called?

**Comment by kpodemski** [comment]
@Hlavtox 

Do I recall correctly that you have resolved this issue?

**Comment by Hlavtox** [comment]
@kpodemski Nope, I needed to use a different approach.

This method is just never called, but its documented everywhere.

**Comment by kpodemski** [comment]
@Hlavtox 

ok, but if I understand correctly, this method is only to indicate whether the cost is external or not, it shouldn't return the actual price

you `return true;` for external cost, `return false;` when you don't have one

I'd have to dig into that, but I'm positive that plenty of modules that I've used did implement handling external shipping costs from API :)

**Comment by Hlavtox** [comment]
@kpodemski Nope, see https://devdocs.prestashop-project.org/9/modules/carrier/#principles

> A carrier module must use the following methods:
> getOrderShippingCost(): to compute the shipping price depending on the ranges that were set in the back office.
> getOrderShippingCostExternal(): to compute the shipping price without using the ranges.

But only the first one works. Why?

```php
if (!$carrier->need_range) {
    return $module->getOrderShippingCostExternal($this);
}
```

Well, so you set `need_range` to false, because you return prices externally right? Makes sense.

Ha. Then the carrier won't be ever offered, because Carrier::getCarriers function doesn't include it. :-)

So you have to use need_range = true, set some nonsense free ranges and use `getOrderShippingCost`.

**Comment by dali-rajab-agillia** [comment]
i'm facing the same problem in Prestashop 9.0.0

**Comment by boo-code** [comment]
Opened #42198 against `9.1.x`.

@Hlavtox to answer the question in your comment - it is not that `getOrderShippingCostExternal()` does not work. The carrier never gets far enough for it to be called.

`Carrier::getCarriersForOrder()` asked for `PS_CARRIERS_AND_CARRIER_MODULES_NEED_RANGE`, which is:

```sql
AND (c.is_module = 0 OR c.need_range = 1)
```

The only group that leaves out is exactly `is_module = 1 AND need_range = 0` - the external carriers. They were removed in SQL, before anything could ask a module for a price. The branch you quoted is downstream of that and has been correct all along.

Two changes were needed, not one. The list now asks for `ALL_CARRIERS`, which adds no clause of its own, so the result is the previous set plus that missing group. And the range availability checks are skipped when `need_range = 0`, because such a carrier has no ranges by definition and `getMaxDeliveryPriceByWeight()`/`ByPrice()` returning false would drop it a second time. The test covers those separately, so neither half can be removed unnoticed.

What does not change: a module returning `false` still removes the carrier a few lines further down, which is the documented way to say it does not serve a destination. So an external carrier is offered when its module prices it and disappears when it declines.

@dali-rajab-agillia this should cover what you hit on 9.0.0 - if you can try the branch against your module I would like to know.
